doc/man/man3/DwmDescriptorIO.hh.3
.TH "DwmDescriptorIO.hh" 3 "libDwm-0.0.20240716" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DwmDescriptorIO.hh \- \fBDwm::DescriptorIO\fP class declaration\&.  

.SH SYNOPSIS
.br
.PP
\fR#include <array>\fP
.br
\fR#include <cassert>\fP
.br
\fR#include <cstdint>\fP
.br
\fR#include <cstdio>\fP
.br
\fR#include <deque>\fP
.br
\fR#include <iostream>\fP
.br
\fR#include <list>\fP
.br
\fR#include <map>\fP
.br
\fR#include <set>\fP
.br
\fR#include <string>\fP
.br
\fR#include <tuple>\fP
.br
\fR#include <unordered_map>\fP
.br
\fR#include <unordered_set>\fP
.br
\fR#include <variant>\fP
.br
\fR#include <vector>\fP
.br
\fR#include 'DwmPortability\&.hh'\fP
.br
\fR#include 'DwmDescriptorIOCapable\&.hh'\fP
.br
\fR#include 'DwmVariantFromIndex\&.hh'\fP
.br

.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBDwm::DescriptorIO\fP"
.br
.RI "This class contains a collection of static functions for reading and writing simple types, in network byte order (MSB first)\&. "
.in -1c
.SH "Detailed Description"
.PP 
\fBDwm::DescriptorIO\fP class declaration\&. 


.PP
\fBAuthor\fP
.RS 4
Daniel W\&. McRobb 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for libDwm-0\&.0\&.20240716 from the source code\&.


ChangeLog
2024-07-25  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as libDwm-0.9.45

	* README.md: mention libfmt dependency for the case where
	std::format is not available.

	* tagged as libDwm-0.9.44

	* classes/include/DwmGZIOCapable.hh: Documentation updates.
	* classes/include/DwmBZ2IOCapable.hh: Documentation updates.
	* classes/include/DwmASIOCapable.hh: Documentation updates.
	* classes/include/DwmDescriptorIOCapable.hh: Documentation updates.
	* classes/include/DwmFileIOCapable.hh: Documentation updates.
	* classes/include/DwmStreamIOCapable.hh: Documentation updates.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmDocMainPage.hh, README.md,
	doc/examples/DwmIOExample3.cc: Documentation updates.

2024-07-24  dwm  <dwm@nomnom.mcplex.net>

	* tagged as libDwm-0.9.43

	* mkrelease.sh: 'push --tags' should be 'push --follow-tags'

	* tagged as libDwm-0.9.42

	* classes/include/DwmSysLogger.hh: Fix pasto on platforms where
	we need libfmt (std:: should be fmt::), namely Raspberry Pi OS.

2024-07-24  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.41

	* classes/tests/TestThreadQueue.cc: Use max concurrency on
	multicore systems when running performance tests, just to
	add to the stress.  Also use C++14 digit separator (') in
	some numeric literals for readability.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/src/DwmSysLogger.cc, classes/include/DwmSysLogger.hh:
	Allow use of std::source_location, and use them in the Syslog
	and FSyslog macros.

	* Removed the following files since I don't need them here:

	classes/include/DwmHostPinger.hh
	classes/include/DwmIpv4IcmpDestinationUnreachable.hh
	classes/include/DwmIpv4IcmpEchoReply.hh
	classes/include/DwmIpv4IcmpEchoRequest.hh
	classes/include/DwmIpv4IcmpMessage.hh
	classes/include/DwmIpv4IcmpRedirect.hh
	classes/include/DwmIpv4IcmpSourceQuench.hh
	classes/include/DwmIpv4IcmpTimeExceeded.hh
	classes/include/DwmIpv4PacketHeader.hh
	classes/include/DwmIpv4TcpHeader.hh
	classes/include/DwmIpv4TcpPacket.hh
	classes/include/DwmIpv4TcpPayload.hh
	classes/include/DwmIpv4UdpHeader.hh
	classes/include/DwmIpv4UdpPacket.hh
	classes/include/DwmIpv4UdpPayload.hh
	classes/include/DwmPatricia.hh
	classes/include/DwmPingDestination.hh
	classes/include/DwmPingDestinationEntry.hh
	classes/include/DwmSocket.hh
	classes/include/DwmXmlAttribute.hh
	classes/include/DwmXmlElement.hh
	classes/src/DwmHostPinger.cc
	classes/src/DwmIpv4IcmpDestinationUnreachable.cc
	classes/src/DwmIpv4IcmpEchoReply.cc
	classes/src/DwmIpv4IcmpEchoRequest.cc
	classes/src/DwmIpv4IcmpMessage.cc
	classes/src/DwmIpv4IcmpRedirect.cc
	classes/src/DwmIpv4IcmpSourceQuench.cc
	classes/src/DwmIpv4IcmpTimeExceeded.cc
	classes/src/DwmIpv4PacketHeader.cc
	classes/src/DwmIpv4TcpHeader.cc
	classes/src/DwmIpv4TcpPacket.cc
	classes/src/DwmIpv4TcpPayload.cc
	classes/src/DwmIpv4UdpHeader.cc
	classes/src/DwmIpv4UdpPacket.cc
	classes/src/DwmIpv4UdpPayload.cc
	classes/src/DwmPingDestination.cc
	classes/src/DwmPingDestinationEntry.cc
	classes/src/DwmSocket.cc
	classes/src/DwmXmlAttribute.cc
	classes/src/DwmXmlElement.cc
	classes/src/Makefile
	classes/tests/Makefile
	classes/tests/TestDwmSocket.cc
	classes/tests/TestHostPinger.cc
	classes/tests/TestIpv4IcmpDestinationUnreachable.cc
	classes/tests/TestIpv4IcmpEchoReply.cc
	classes/tests/TestIpv4IcmpEchoRequest.cc
	classes/tests/TestIpv4IcmpRedirect.cc
	classes/tests/TestIpv4IcmpSourceQuench.cc
	classes/tests/TestIpv4IcmpTimeExceeded.cc
	classes/tests/TestIpv4PacketHeader.cc
	classes/tests/TestIpv4TcpHeader.cc
	classes/tests/TestIpv4TcpPacket.cc
	classes/tests/TestIpv4UdpHeader.cc
	classes/tests/TestIpv4UdpPacket.cc
	classes/tests/TestPatricia.cc
	classes/tests/ip.pcap
	classes/tests/tcp.pcap
	classes/tests/udp.pcap

2024-07-23  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* getvers.sh: bump shared library version to reflect ABI changes
	(vtables changed from removal of inheritance).

	* README.md: Document Read and Write requirements for
	ASIO with boost::asio::generic::stream_protocol::socket

	* classes/include/DwmIpv4IcmpMessage.hh,
	  classes/include/DwmIpv4PacketHeader.hh,
	  classes/include/DwmIpv4TcpHeader.hh,
	  classes/include/DwmIpv4TcpPacket.hh,
	  classes/include/DwmIpv4TcpPayload.hh,
	  classes/include/DwmIpv4UdpHeader.hh,
	  classes/include/DwmIpv4UdpPacket.hh,
	  classes/include/DwmIpv4UdpPayload.hh:
	No longer need to inherit from *Readable and *Writable classes.

	* classes/include/DwmASIOCapable.hh,
	  classes/include/DwmBZ2IOCapable.hh,
	  classes/include/DwmDescriptorIOCapable.hh,
	  classes/include/DwmFileIOCapable.hh,
	  classes/include/DwmGZIOCapable.hh,
	  classes/include/DwmStreamIOCapable.hh:
	Minor cleanup to concepts.

	* tagged as libDwm-0.9.40

	* doc/.gitignore: ignore libDwm.tag (generated by doxygen)

	* README.md: Add mkfbsdmnfst and mkdebcontrol as build
	dependencies.

2024-07-22  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Removed IN-PROGRESS file.

	* tagged as libDwm-0.9.39

	* Added README.md for the github mirror.

2024-07-21  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.38

2024-07-21  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* doc/footer.html: Tweak to point at my web site.

	* classes/src/DwmMacAddress.cc: Use sizeof() where appropriate.

	* Remove doxygen-generated documentation from repository; it's
	build output that doesn't need to be in the repository.

2024-07-20  Daniel McRobb  <dwm@toto.mcplex.net>

	* tagged as libDwm-0.9.37

2024-07-20  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* classes/src/DwmProcessTableLex.lex,
	classes/src/DwmProcessTableParse.y: 'ps' output on FreeBSD shows
	a single hyphen ('-') instead of a ctime() string for lstart for
	some processes.  Patch this by substituting the lstart from the
	nearest ancestor of the process that has a valid lstart.

	* classes/include/DwmBootTime.hh, classes/include/DwmUptime.hh
	classes/src/DwmBootTime.cc, classes/src/DwmUptime.cc:

classes/src/DwmDescriptorIO.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2004-2007, 2013, 2017, 2020
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR
//  PURPOSE, OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY 
//  PATENT, TRADEMARK OR OTHER RIGHTS.
//===========================================================================
//---------------------------------------------------------------------------
//!  \file DwmDescriptorIO.cc
//!  \author Daniel W. McRobb
//!  \brief Dwm::DescriptorIO class implementation
//---------------------------------------------------------------------------

extern "C" {
  #include <sys/types.h>
  #include <unistd.h>
  #include <fcntl.h>
}

#include <cstdlib>
#include <cassert>

#include "DwmSysLogger.hh"
#include "DwmDescriptorIO.hh"
#include "DwmPortability.hh"
#include "DwmXDRUtils.hh"

namespace Dwm {

  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, char c)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      rc = write(fd,(void *)&c,sizeof(c));
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, uint8_t c)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      rc = write(fd, (void *)&c, sizeof(c));
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, bool b)
  {
    uint8_t  c = b;
    return(DescriptorIO::Write(fd, c));
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, int16_t val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      int16_t v = htobe16(val);
      rc = write(fd, (void *)&v, sizeof(v));
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, uint16_t val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint16_t v = htobe16(val);
      rc = write(fd, (void *)&v, sizeof(v));
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, int32_t val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      int32_t v = htobe32(val);
      rc = write(fd, (void *)&v, sizeof(v));
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, uint32_t val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint32_t v = htobe32(val);
      rc = write(fd, (void *)&v, sizeof(v));
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, const int64_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      int64_t v = htobe64(val);
      rc = write(fd, (void *)&v, sizeof(v));
      if (rc != sizeof(v)) {
        rc = -1;
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, const uint64_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint64_t v = htobe64(val);
      rc = write(fd, (void *)&v, sizeof(v));
      if (rc != sizeof(v)) {
        rc = -1;
      }
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, float val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      char     *buf = 0;
      uint32_t  xe = XDRUtils::Encode(val, &buf);
      rc = write(fd, buf, xe);
      free(buf);
      if (rc < 4) {
        rc = -1;
      }
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, const double & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      char  *buf = 0;
      uint32_t  xe = XDRUtils::Encode(val, &buf);
      rc = write(fd, buf, xe);
      free(buf);
      if (rc < 8) {
        rc = -1;
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, const std::string & s)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint64_t  len = s.size();
      if (Write(fd,len) == sizeof(len)) {
        rc = sizeof(len);
        if (len > 0) {
          if (Write(fd, (const void *)s.c_str(), len) == len) {
            rc += len;
          }
          else {
            rc = -1;
          }
        }
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, char & c)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      rc = read(fd, (void *)&c, sizeof(c));
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, uint8_t & c)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      rc = read(fd, (void *)&c, sizeof(c));
    }
    return(rc);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, bool & b)
  {
    uint8_t  c;
    ssize_t  rc = DescriptorIO::Read(fd, c);
    if (rc > 0) {
      b = c ? true : false;
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, int16_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      int16_t  v;
      rc = read(fd, (void *)&v, sizeof(v));
      if (rc == sizeof(v)) {
        val = be16toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, uint16_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint16_t  v;
      rc = read(fd, (void *)&v, sizeof(v));
      if (rc == sizeof(v)) {
        val = be16toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, int32_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      int32_t  v;
      rc = read(fd, (void *)&v, sizeof(v));
      if (rc == sizeof(v)) {
        val = be32toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, uint32_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint32_t  v;
      rc = read(fd, (void *)&v, sizeof(v));
      if (rc == sizeof(v)) {
        val = be32toh(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, int64_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint64_t  v;
      if (read(fd, (void *)&v, sizeof(v)) == sizeof(v)) {
        val = be64toh(v);
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, uint64_t & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint64_t  v;
      if (read(fd, (void *)&v, sizeof(v)) == sizeof(v)) {
        val = be64toh(v);
        rc = sizeof(v);
      }
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, float & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      char     buf[4];
      rc = read(fd, buf, 4);
      if (rc == 4) {
        XDRUtils::Decode(buf, val);
      }
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, double & val)
  {
    ssize_t  rc = -1;
    if (fd >= 0) {
      char     buf[8];
      rc = read(fd, buf, 8);
      if (rc == 8) {
        XDRUtils::Decode(buf, val);
      }
    }
    return(rc);
  }
  
  //------------------------------------------------------------------------
  //!  
  //------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, std::string & s)
  {
    s.clear();
    ssize_t  rc = -1;
    if (fd >= 0) {
      uint64_t  len;
      if (Read(fd, len) == sizeof(len)) {
        rc = sizeof(len);
        if (len > 0) {
          try {
            s.resize(len);
            ssize_t  bytesRead = Read(fd, s.data(), len);
            if (bytesRead == len) {
              rc += len;
            }
            else {
              s.clear();
              rc = -1;
            }
          }
          catch (const std::exception & ex) {
            Syslog(LOG_ERR, "Exception: %s", ex.what());
            rc = -1;
          }
          catch (...) {
            Syslog(LOG_ERR, "Exception");
            rc = -1;
          }
        }
      }
    }
    return(rc);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Read(int fd, void *buf, size_t buflen)
  {
    ssize_t  totalBytesRead = 0;
    if (buf && (buflen > 0) && (fd >= 0)) {
      uint8_t  *bufp = (uint8_t *)buf;
      while (totalBytesRead < buflen) {
        ssize_t  bytesRead = read(fd, (void *)bufp, buflen - totalBytesRead);
        if (bytesRead == buflen) {
          totalBytesRead = bytesRead;
          break;
        }
        else if (bytesRead <= 0) {
          break;
        }
        else {
          bufp += bytesRead;
          totalBytesRead += bytesRead;
        }
      }
    }
    return ((totalBytesRead == buflen) ? buflen : -1);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  ssize_t DescriptorIO::Write(int fd, const void *buf, size_t buflen)
  {
    ssize_t  totalBytesWritten = 0;
    if (buf && (buflen > 0) && (fd >= 0)) {
      const uint8_t  *bufp = (const uint8_t *)buf;
      while (totalBytesWritten < buflen) {
        ssize_t  bytesWritten = write(fd, (void *)bufp,
                                      buflen - totalBytesWritten);
        if (bytesWritten == buflen) {
          totalBytesWritten = bytesWritten;
          break;
        }
        else if (bytesWritten < 0) {
          break;
        }
        else {
          bufp += bytesWritten;
          totalBytesWritten += bytesWritten;
        }
      }
    }
    return ((totalBytesWritten == buflen) ? buflen : -1);
  }
  

}  // namespace Dwm


//---------------------------- emacs settings -----------------------------
//  Local Variables:
//  mode: C++
//  tab-width: 2
//  indent-tabs-mode: nil
//  c-basic-offset: 2
//  End:
//-------------------------------------------------------------------------
